Camera platform for buienradar imagery

ties opened the pull request:
## Description:
This PR adds a [dutch rain radar](http://www.buienradar.nl) camera component that delivers the weather-radar imagery from [buienradar.nl](http://www.buienradar.nl) to the existing `buienradar` integration. 

This idea was inspired by https://github.com/home-assistant/home-assistant/pull/22816. _Most_ of the current PR's functionality could be provided by a generic IP camera using the given URL from `buienradar.nl` documentation.

The main contribution of this PR is that it caches images for the given `delta` and prevents multiple parallel http calls, which happened with multiple browser tabs open.

This behaviour is verified in the test. When testing the PR manually this situation can be exacerbate by adding an `asyncio.Sleep` in the function that makes the actual http call. This behaviour is different from `util.Throttle` because it does not return the old value but instead it waits for the request to finish. The generic IP does not include this feature.

In addition it adds the camera component by implicitly documenting the URL. Factoring out the http request to a PyPI package would create a trivial package which would hardly be re-usable since it needs aiohttp passed in and would return `(result_data: Optional[bytes], result_status: bool, last_modified: Optional[str])` triples instead of a more idiomatic python api which would return just the result.

Alternatively the improved re-entrant behaviour could be added to the Generic IP camera and the radar imagery URL documented in the `buienradar` component.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here> to be added after discussion on the PR's utility.

## Example entry for `configuration.yaml` (if applicable):
```yaml
camera:
  - platform: buienradar
    name: buienradar
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [ ] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.


[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

ties committed:
Add camera for buienradar radar

ties committed:
Use asyncio.Conditions instead of asyncio.Lock

ties committed:
Add test and fix python 3.5 compatibility

ties committed:
rename interval to delta for consistency with BOM integration

homeassistant commented:
Hi @ties, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#23358).

Once you do that we will be able to review and accept this pull request.

Thanks!

ties committed:
fix linting error introduced during rebase

balloob commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,176 @@
+CONF_DIMENSION = 'dimension'
+CONF_DELTA = 'delta'
+
+RADAR_MAP_URL_TEMPLATE = ('https://api.buienradar.nl/image/1.0/'
```
Can we add the constant to the existing PyPI package?

balloob replied:
Last comment is regarding this. Would be great if we could just do something like `url = radar_url(x, y)`

balloob replied:
And import `radar_url` from the dependency.

ties replied:
Contributed this to the library and awaiting the result of a pull request upstream. Will update code of the this PR + the buienradar sensor afterwards

balloob commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,176 @@
+        return self._name
+
+    # Cargo-cult from components/proxy/camera.py
+    def camera_image(self) -> Optional[bytes]:
```
We should not need this method as we always call `async_camera_image` instead.

balloob commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,176 @@
+
+        return dt_util.utcnow() > self._deadline
+
+    async def __retrieve_radar_image(self) -> bool:
```
Convention is that for private methods, use a single underscore prefix. Double underscore prefix is reserved for Python built-in methods.

ties replied:
I used double underscores since [name mangling](https://docs.python.org/3.7/tutorial/classes.html#private-variables) prevents conflicts with function names from parent classes.

This is different from `__[builtin]__`which has two trailing underscores. I will adjust this for `__retrieve_radar_image` but would prefer to keep it for the other case if mangled names are allowed in this project's style.

balloob replied:
Oh, TIL.

balloob commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,176 @@
+            return self._last_image
+        finally:
+            # get lock, unset loading status, notify all waiting tasks
+            async with self._condition:
```
If one request is waiting for the result, it does so inside the `async with self._condition`. Doesn't that mean it never releases the lock ? And if so, we will never be able to acquire the lock here?

ties replied:
There could be deadlocks in here if I missed something (especially with exceptions) but this is not one 🙂

`await self._condition.wait()` releases the lock and acquires it before continuing. I updated the docstring to clarify this

balloob replied:
Oh, TIL.

balloob commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,176 @@
+        """
+        Return a still image response from the camera.
+
+        Uses ayncio conditions to make sure only one task enters the critical
```
I wonder if we need to make this a decorator that can be reused. We have similar logic inside Hue and probably will have it for other integrations that share a resource among requests.

ties replied:
I considered this and will get back on this later (likely this weekend).

I see three scenario's here:
  * generator function used as coroutine, [seems](https://stackoverflow.com/questions/34469060/python-native-coroutines-and-send) to rule out async functions. I do not find this intuitive
  * a decorator. Advantage: does not depend on function name, does not need to be called explicitly. Tricky because it will take parameters (the time delta) from the instance of the class and this protocol is implicit.
  * an abstract base class with multiple inheritance. Advantage: Makes the lifecycle explicit. Disadvantage: more verbose

ties replied:
I think I have found the generic/re-usable use case, will try to write down this design pattern later this week due to time constraints.

In the meantime I have made an assumption about lifecycle of integrations. The state returned is the state _at the time of the return_, instead of the time of calling? This would allow calls to be '"[folded](https://en.wikipedia.org/wiki/Fold_(higher-order_function)" to single calls in my design pattern.

This pattern can combine multiple "request update" calls into one, or combine multiple light switch actions into a single action performed on an integration (`turn on A`, `turn on B`, `turn on C` => `turn on (A, B, C,)`)

balloob replied:
State is at time of the return.

We don't do folded turn_on yet by platform. Right now we do allow platforms to specify max concurrency for calling services. States are only polled once all commands have been sent.

ties committed:
Improved buienradar.camera documentation and tests

  * Incorporated one comment on a redundant/cargo cult function
  * Improved documentation
  * Increase test coverage by fixing one test by making it a coroutine
    (to make it actually run), adding another test case, and changing
    the flow in the implementation.

MartinHjelmare commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,178 @@
+                               discovery_info=None):
+    """Set up buienradar radar-loop camera component."""
+    c_id = config.get(CONF_ID)
+    dimension = config.get(CONF_DIMENSION) or 512
```
Please move all config defaults to the config schema.

MartinHjelmare commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,178 @@
+                 delta: float):
+        """Initialize the component."""
+        super().__init__()
+        self._hass = hass
```
`hass` will be set on the entity, as `self.hass`, when the entity has been added to home assistant.

MartinHjelmare commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -0,0 +1,178 @@
+        self._dimension = dimension
+        self._delta = delta
+
+        self._condition = asyncio.Condition(loop=hass.loop)
```
If we create the condition before creating the entity, and pass the condition to the entity, we don't need to pass in `hass` to the entity.

We don't need to set `loop` explicitly on the condition if we're in a coroutine.

ties replied:
Creating a condition before creating the entity obscures the invariant that the condition is private to this entity. I would like keep the condition object internal because of this (sharing it would cause `notify_all` to notify where it shouldn't) and will document that.

I agree that it is not needed to store a private reference to `hass`, will add an optional parameter for the event loop since it is considered good practice to do so.

You imply an invariant here which was not clear to me from the documentation and other code. Note that this invariant is logical but does not necessarily implicitly hold.

I think that this holds: "async setup methods are called with the event loop of the thread (i.e. `asyncio.get_event_loop()` being equal to hass.loop". If this holds it would really help if part of the platform lifecycle  is documented :)

MartinHjelmare replied:
I'm not 💯 on what you mean with invariant. Sorry, I don't have a computer science background.

> You imply an invariant here which was not clear to me from the documentation and other code. Note that this invariant is logical but does not necessarily implicitly hold.

What are you referring to here? `hass` being set on the entity on addition to home assistant or current event loop being used for `loop` by default inside coroutines (inside the event loop)?

MartinHjelmare replied:
We can create the condition inside init without passing in hass if we don't set loop explicitly. That's my suggestion after your explanation about condition being private, which makes sense.

ties replied:
My explanation was not 100% precise either, I removed an extraneous "implicitly". With invariant I mean "a property that always holds"/"it always is true that ..."

> What are you referring to here? hass being set on the entity on addition to home assistant or
> current event loop being used for loop by default inside coroutines (inside the event loop)?

To me it was not clear that `hass.loop` (in `async_setup_component`) is the same as the default event loop used by coroutines (inside the event loop). It feels logical but would cause potentially cause bugs if it is not. Being unsure I would prefer to pass it in explicitly

MartinHjelmare replied:
The default loop in coroutines is the current loop used in the event loop that runs the coroutines. This is part of asyncio library. Since we're in a coroutine in `async_setup_platform` we don't need to care about what loop to use. We just ask asyncio to use the current loop by not setting it explicitly. This is safe.

ties replied:
> The default loop in coroutines is the current loop used in the event loop that runs the coroutines. This is part of asyncio library. Since we're in a coroutine in `async_setup_platform` we don't need to care about what loop to use. We just ask asyncio to use the current loop by not setting it explicitly. This is safe.

There is a tiny, tiny caveat there. From my understanding, `asyncio.get_running_loop` gets the current loop. By default asyncio Lock's/Condiotion's/... [use](https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L162) `asyncio.get_event_loop()`.

This _can be_ a different one, and will cause "surprising" issues in that situation (RuntimeError's or difficult to debug situations when there are multiple event loops)... Glad that is not the case here.

MartinHjelmare replied:
The [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop) on `asyncio.get_event_loop` doesn't say it can return a different loop than the current one, if I'm reading correctly. If there's no current loop set, it will create a new loop and set it as current. Yeah, there might be cases where this can be a problem. I don't see cases at the moment that I'm afraid should impact us.

MartinHjelmare replied:
There's this note in the docs though, just below this [section](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.new_event_loop):
> Note that the behaviour of get_event_loop(), set_event_loop(), and new_event_loop() functions can be altered by setting a custom event loop policy.

ties replied:
> Yeah, there might be cases where this can be a problem. I don't see cases at the moment that I'm afraid should impact us.

Exactly! It is one of those 'should not happens' that cause people to be careful. I did a quick grep for this yesterday and the only place in this project (note that libraries can break this...) that could cause issues that I found is usage of `homeassistant.util.async.asyncio_run` which would unset the event loop afterwards.

But in coroutines it still seems to work in practice. See https://gist.github.com/ties/21d75aa31c399e6c9a100232d040f214 for a bunch of test cases :)

> There's this note in the docs though, just below this [section](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.new_event_loop):
> 
> > Note that the behaviour of get_event_loop(), set_event_loop(), and new_event_loop() functions can be altered by setting a custom event loop policy.

The general gist of this is what caused me to be defensive: As a new contributor you see that event loops are commonly passed in explicitly in the project and there are few guarantees.

I'll keep a note that it may be good to add the fact that the event loop returned by `get_event_loop` is assumed to be the correct one to the documentation. Without a custom event loop this is actually what seems to happen.

balloob replied:
They fixed it in Python 3.5.3 so that get_event_loop called from inside an async function will always return the event loop that is running the async task: https://bugs.python.org/issue28613

(see also https://docs.python.org/3.5/whatsnew/changelog.html#python-3-5-3 and search for get_event_loop)

balloob replied:
If you see code passing in the event loop, it's because it was most likely written prior to our minimum requirement being raised to 3.5.3

ties replied:
Really nice piece of clarification. Thanks! 👍

MartinHjelmare commented on tests/components/buienradar/test_camera.py:
```diff
@@ -0,0 +1,172 @@
+            "image/1.0/RadarMapNL?w={dim}&h={dim}").format(dim=dim)
+
+
+@asyncio.coroutine
```
Use `async def` and `await` syntax.

MartinHjelmare commented on tests/components/buienradar/test_camera.py:
```diff
@@ -0,0 +1,172 @@
+    """Test that it fetches the given url."""
+    aioclient_mock.get(radar_map_url(), text='hello world')
+
+    yield from async_setup_component(hass, 'camera', {
```
See above.

ties replied:
Cargo cult copied this. Prefer this style and will change it.

ties committed:
style changes after review, additional test case

MartinHjelmare commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -60,7 +60,13 @@ class BuienradarCam(Camera):
-    _delta = 0.0         # type: float
+    _delta = 600.0          # type: float
+    """
+    Ensures that only one reader can cause an http request at the same time,
```
This should be a comment instead, using `#`.

ties replied:
I thought docstrings need the `"""` style. I just found out that docstrings should be below the variable though in many style guides (e.g. [google](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)). Will move this one and the one on line 60 when I get to it (likely: tomorrow night)

MartinHjelmare replied:
Yes docstring need `"""` and have to be put right under the class or function. Variables don't have their own docstrings.

ties replied:
They seem to have their own docstrings in the styleguide (see below). On the other hand, sphinx/pydoc do not seem to pick this up so I'm fine with changing it to a comment.
```py
    def __init__(self, param1, param2, param3):
# ...
        self.attr3 = param3  #: Doc comment *inline* with attribute

        #: list of str: Doc comment *before* attribute, with type specified
        self.attr4 = ['attr4']

        self.attr5 = None
        """str: Docstring *after* attribute, with type specified."""
```

ties replied:
Switched style since I do not hold a strong preference one way or the other and the style of comment is looser: No first line summary + whitespace + body required..

ties commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -60,7 +60,13 @@ class BuienradarCam(Camera):
     _deadline = None        # type: Optional[datetime]
     _name = ""              # type: str
-    _delta = 0.0         # type: float
+    _delta = 600.0          # type: float
```
I do not know how I can add type annotations while supporting Python 3.5 (without [PEP 526](https://www.python.org/dev/peps/pep-0526/)) support otherwise.

```py
class Foo(object):
  bar  # type: int
```
Gives a `NameError: name 'bar' is not defined`

What would you propose?

MartinHjelmare replied:
It should work in init method.

MartinHjelmare replied:
Does this work? (I'm on mobile at the moment)
```py
class Foo:

    def __init__(self):
        self.bar = None  # type: int
```

ties replied:
"Kind of works": mypy detects the types and has editor support.

The `self.bar = 0  # type: int` style does not create `module.__annotations__`. Nor does `self.bar: int = 0`. But functionally they work: mypy gives errors if you assign a different type.

mypy also infers types, `self.bar = 2` becomes an int, `self.bar = untyped_dict['a']` does not and needs the annotation.

Since we are on Python 3.5 for now I think the goal is tool support thus moving to types in the constructor is a good alternative.

With Python 3.6+ the assignment is no longer needed and the annotations become viable. I wish the annotations were just stripped on build instead of having to use the legacy style, I prefer the explicit annotations by now.

ties commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -29,22 +29,22 @@
-    dimension = config.get(CONF_DIMENSION) or 512
-    delta = config.get(CONF_DELTA) or 600.0
-    name = config.get(CONF_NAME) or "Buienradar loop"
+    dimension = config.get(CONF_DIMENSION)
```
Can be changed after the earlier change, thanks!

ties committed:
Use python 3.5 style mypy type annotations in __init__

ties commented:
I think for as far as the changes are up to me this PR is done. What is left is removing the `radar_url` utility after the url's are merged upstream, and a final rebase + squash.

An optional feature is configurable image size (`x` _and_ `y` instead of one dimension) but I'm personally fine with the square image and I prefer it. Can add this if other `Camera` implementations have this.

Let's give upstream https://github.com/mjj4791/python-buienradar/pull/7 some time to be merged…

ties committed:
Remove explicit passing of event loop

rytilahti commented:
Any updates on this?

Also, on a related matter (being very, very offtopic here, apologies): the buienalarm on Android (I suppose it uses the same data source?) is the best weather app I have used. The graph (as implemented here to my understanding, to see when it's safe to go out again) and especially it's alarms ("It starts to rain in 25 minutes and continues for 40 minutes") are awesome. So my question  is, would it be possible to extend this further to include the alarms (probably in a separate PR, though)?

ties commented:
Pull request is still up to date. However upstream has still not responed to my pull request and I think it's not worth it to fork or create a package purely to have an URL constant.

ties committed:
Adopt buienradar camera as codeowner

balloob committed:
Update manifest.json

balloob commented:
Please run `python3 -m script.hassfest` and commit the results.

ties committed:
Update CODEOWNERS through hassfest

Updated CODEOWNERS through hassfest (instead of manually), thanks to
@balloob for the hint.

tjorim reviewed (commented):
Documentation is missing.

mjj4791 commented:
@ties 
I was trying to configure this component yesterday, found out that there is no documentation to be found online (as mentioned before/above); are you planning to add it, so other users can find your component as well?

MartinHjelmare commented:
Please open an issue in our docs repo.

Merged PRs should not be used for enhancement discussion or bug reports.

Thanks!
